typing/includemod.ml

```diff
@@ -767,36 +781,41 @@ and signature_components  ~in_eq ~loc old_env ~mark env subst
             let shape_map =
               Shape.Map.add_class_proj shape_map id1 orig_shape
             in
-            let item = unrecoverable_error item in
+            let item = mark_error_as_unrecoverable item in
             id1, item, shape_map, true
         | Sig_class_type(id1, info1, _, _), Sig_class_type(_id2, info2, _, _) ->
             let item =
               class_type_declarations ~loc ~old_env env subst info1 info2
             in
+            let item = mark_error_as_unrecoverable item in
             let shape_map =
               Shape.Map.add_class_type_proj shape_map id1 orig_shape
             in
-            let item = unrecoverable_error item in
             id1, item, shape_map, false
         | _ ->
             assert false
       in
-      let oks, (final_map, deep_modifications), errors, unknowns as r =
+      let r =
         match item with
         | Error { recoverable = false; error } ->
-            let final_map = shape_map, !shape_modified in
-            [], final_map, [id,error], rem
+            { runtime_coercions=[];
+              shape_map;
+              deep_modifications= !shape_modified;
+              errors=[id,error];
+              leftovers=rem
+            }
         | Error { recoverable = true; _ } | Ok _ ->
             signature_components ~in_eq ~loc old_env ~mark env subst
               orig_shape shape_map rem
       in
-      let final_map = final_map, deep_modifications || !shape_modified in
+      let deep_modifications = r.deep_modifications || !shape_modified in
       match item with
       | Ok x when present_at_runtime ->
-          (pos,x) :: oks, final_map, errors, unknowns
-      | Ok _ -> oks, final_map, errors, unknowns
+          let runtime_coercions = (pos,x) :: r.runtime_coercions in
+          { r with runtime_coercions; deep_modifications }
+      | Ok _ -> { r with deep_modifications }
       | Error {error; recoverable = true }  ->
-          oks, final_map, (id,error) :: errors, unknowns
+          { r with errors = (id,error) :: r.errors; deep_modifications }
       | Error {recoverable=false; _ }  -> r
```

Comment:
The code flow is weird here, in particular the flow of `deep_modifications`. A proposal:

```ocaml
let first = (* signature difference for the first item *)
  match item with
  | Ok -> ...
  | Error -> ...
in
let continue =
  match item with
  | Ok _ -> true
  | Error e -> e.recoverable
in
let rest = (* signature difference for the remainder *)
  if not continue then ...
  else signature_items ...
in
{
  runtime_coercions = first.runtime_coercions @ rest.runtime_coercions;
  shape_map =
    (* shape maps are threaded through the computation, the last one has all information *)
    rest.shape_map;
  deep_modifications = first.deep_modifications || rest.deep_modifications;
  errors = first.errors @ rest.errors;
  leftovers = first.leftovers @ rest.leftovers;
}
```

Reply:
Going in this direction, I would propose to move helper functions in a submodule and go for something like:
```ocaml
let deep_modifications = !shape_modified in
let first =
  match item with
  | Ok x ->
    let runtime_coercions = if present_at_runtime then [pos,x] else [] in
    Sign_diff.make ~deep_modifications ~runtime_coercions shape_map
  | Error { recoverable = true; error } ->
    Sign_diff.make ~errors:[id,error] ~deep_modifications shape_map
  | Error { recoverable = false; error } ->
    Sign_diff.make ~deep_modifications ~errors:[id,error] ~leftovers:rem
      shape_map
   in
   let continue = match item with
   | Ok _ -> true
   | Error x -> x.recoverable
   in
   let rest =
     if continue then
       signature_components ~in_eq ~loc old_env ~mark env subst
         orig_shape shape_map rem
     else Sign_diff.empty
   in
   Sign_diff.merge first rest
  ```

Reply:
This looks fine. (I'm not fond of turning records into functions with labelled and optional parameters. I would expose the records directly, possibly with `{ empty with ...}` for initialization. But this is just me.)

Reply:
@gasche : last check before squashing and merging, are you fine with the last commit?